### PR TITLE
Feature: Improve installation and setup

### PR DESCRIPTION
**Pull request by pke**
The plugin now automatically setting up all the manifest and URL scheme changes mentioned to do manually in AndroidManifest.xml and xcode projects.

The Android part adds an InsertApplication to cope with the maximum DEXable function numbers.

This also adds some API documentation.

### CODE REVIEW

**Comment by night2tl** [comment]
Thanks for the PR Philipp, we're reviewing it. Much appreciated!

**Comment by pke** [comment]
Just noticed, that I did not finish the `init` JS function. It should allow the specification of userAttributes and userId and call those functions. But I guess I'll remove this functionality for now.

**Comment by night2tl** [comment]
Hey Philipp, we've been a bit busy in the past few days, but I see you've made some more changes, I'll go over this as soon as I can. Again, thanks for all the help and for the PR. Really appreciated.

**Comment by night2tl** [comment]
Hey Phillip, I can't seem to call init from the js, although the plugin exports it, I get an exception saying I should use cordova threadpool, but when I use it, it doesn't fix the problem. Any ideas concerning the issue? If we can't resolve it, we'll have to add the initialization in the native part manually.

File: README.md [hunk 1]
```diff
@@ -3,37 +3,102 @@
...
+You can ignore all the setup instructions in the SDK Integration document, since this plugin does handle all that for 
+you automatically.
+
+## Collect required variables for plugin installation
+
+### All supported platforms need
+
+#### `COMPANY_NAME`
+
+This is equal to your `companyname.insert.io` subdomain but you can also find it in the SDK Integrations document under "Step 3" in the `initSDK` code call.
```
Comment (night2tl) [inline README.md:27]: you can find all of those in the app details section as well

**Comment by pke** [comment]
Thanks for reviewing @night2tl 
I didn't have problems calling init from the JS side. I have currently a testflight app out there for our customer to test the code.

Ah, ok you are testing the Android version of the plugin. I'll have to look into that again later.

**Comment by night2tl** [comment]
Thanks, I'll be waiting

File: README.md (hunk 1 above)
Comment (pke) [reply README.md:27]: Right, I'll update the docs.

**Comment by pke** [comment]
@night2tl Strange. The iOS code works just fine. However on Android, even inside a Runnable I hit the error callback. Unfortunately your SDK does not give any information whatsoever **why** the init call failed. How about adding an error description argument to the `InsertPhasesCallbackInterface`?

**Comment by night2tl** [comment]
@pke this is what I get. I also put it in a runnable like so:

this.cordova.getThreadPool().execute(new Runnable() {
                @Override
                public void run() {
...
});

_______________________________________________________________________________
E/AndroidRuntime(30726): java.lang.ExceptionInInitializerError
E/AndroidRuntime(30726): 	at io.insert.cordova.plugin.Insert$1.run(Insert.java:44)
E/AndroidRuntime(30726): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
E/AndroidRuntime(30726): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
E/AndroidRuntime(30726): 	at java.lang.Thread.run(Thread.java:818)
E/AndroidRuntime(30726): Caused by: java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare()
E/AndroidRuntime(30726): 	at android.os.Handler.<init>(Handler.java:200)
E/AndroidRuntime(30726): 	at android.os.Handler.<init>(Handler.java:114)
E/AndroidRuntime(30726): 	at sdk.insert.io.Insert.<clinit>(Unknown Source)
E/AndroidRuntime(30726): 	... 4 more

**Comment by pke** [comment]
Try `activity.runOnUiThread`. It seems your SDK wants to perform some UI thread related things.

**Comment by pke** [comment]
@night2tl I tried to run `init` in the UI thread now but it times out after 10 (?) seconds.

**Comment by night2tl** [comment]
@pke I get the same. Pretty weird, because if I just run it in java from an Application object it works fine. Any other suggestions?

**Comment by pke** [comment]
@night2tl Since I only have the SDK without sources, I can't debug any further. Your side needs to check that. To me it looks like a lock/semaphore thing.
Running from the Application object is really not an option if the plugin should be versatile. Cordova apps could bring their own Application object, which would have to be modified to use the Insert lib.
Also in the Application class (if the plugin would provide it) one could not access the cordova config.xml preferences where usually things like company name and appkey are to be configured.

You should really check whats going on inside the SDK that causes this weird behaviour.

**Comment by night2tl** [comment]
Hey Phillip, 
we fixed the problems with the SDK.
Feel free to use the side branch we have on our repository. (It contains your changes as well)
This is a debug version only for testing

**Comment by pke** [comment]
Great news. Will this also work with the init phase callback?
So this version can't be used in a release build you say?

**Comment by night2tl** [comment]
@pke It can't be used in release build, let us know how it works for you. The init phase callback will also work. 
Give us a heads up once it works well for you, and we'll release a release build.

**Comment by pke** [comment]
@night2tl I am getting this console output now:

```
04-19 18:18:47.481 8561-8643/de.evopark.axa W/PluginManager: THREAD WARNING: exec() call to Insert.init blocked the main thread for 87ms. Plugin should use CordovaInterface.getThreadPool().
04-19 18:20:19.299 8561-8561/de.evopark.axa W/System.err: SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
04-19 18:20:19.299 8561-8561/de.evopark.axa W/System.err: SLF4J: Defaulting to no-operation (NOP) logger implementation
04-19 18:20:19.299 8561-8561/de.evopark.axa W/System.err: SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
04-19 18:20:20.833 8561-8703/de.evopark.axa I/Insert: Insert SDK was successfully integrated and connected to the Insert backend. You can start adding inserts to your app in the Insert's web site. App version identified: '5.3.1'.
```

Do you know about the SLF4J errors?

So it seems to work, basically. I have to reconfigure my testapp to use my test credentials and the init phase callback to see if it fully works.

When you already make changes in the SDK would it be possible to give the fail init phase callback an argument with an Exception object? That would make error reporting in the app easier.

**Comment by night2tl** [comment]
@pke unfortunately we cannot add this at this moment, as it's a big change as other customers will have to comply with it.

**Comment by pke** [comment]
> unfortunately we cannot add this at this moment, as it's a big change as other customers will have to comply with it.

@night2tl I guess you are referring to the init phase callback change. Well, since its a new version of the SDK anyway you can introduce breaking changes without affecting other customers. 
A sane way to bring in the new functionality in an evolutionary way and without breaking would be like this:

```java
interface InsertPhasesCallbackInterface2 extends InsertPhasesCallbackInterface {
  public void onInitFailed(Exception e);
}
```

Then in your SDK code when you want to call the callback you do a dynamic cast like this:

```java
if (callback instanceof InsertPhasesCallbackInterface2) {
  ((InsertPhasesCallbackInterface2)callback).onInitFailed(exception);
} else {
  callback.onInitFailed();
}
```

That combined with a major version bump according to semver should be fine and not affect other customers.

What you think?

**Comment by nirsegev** [comment]
Hi Phillip,
Unfortunately, at this point of time, we do not plan to add this exception argument to the interface. 
We will follow up if it will be added in the future.
Also, I would like to schedule a meeting together with our product manager to discuss your audience requirements. will send you an invite.
Best,
Nir

**Comment by pke** [comment]
Ok, your call. Anyway this branch seems to work, I can init the SDK.

**Comment by pke** [comment]
Hmmm I can't see the app version in the Insert dashboard. I also did not get a confirmation message that a new version was connected.

**Comment by night2tl** [comment]
Hey Phillip, 
for further inspection, you can add the following line in your manifest:

 \<meta-data android:name="insert_debug_logging" android:value="true" /\>

and "adb logcat| grep okhttp" to see where the problem you're experiencing is.

**Comment by pke** [comment]
I am getting 
`04-20 13:18:34.185 27516-27816/de.evopark.axa I/Insert: Insert SDK was successfully integrated and connected to the Insert backend. You can start adding inserts to your app in the Insert's web site. App version identified: '5.3.1'.` in the log but no `okhttp` messages.
Also in the dashboard no 5.3.1 is shown for the android app.

**Comment by pke** [comment]
uhmmm @night2tl this was tested by your side? Cause I had some more things pending in my local branch.

**Pull request merged**